src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs

```diff
@@ -332,77 +333,129 @@ public static unsafe void ClearWithoutReferences(ref byte b, nuint byteLength)
+                if (!Vector.IsHardwareAccelerated || Vector<byte>.Count / sizeof(IntPtr) > 8) // JIT turns this into constant true or false
+                {
+                    if (pointerSizeLength >= 8)
+                    {
+                        nuint stopLoopAtOffset = pointerSizeLength & ~(nuint)7;
+                        do
+                        {
+                            Unsafe.Add(ref ip, (nint)i + 0) = default;
+                            Unsafe.Add(ref ip, (nint)i + 1) = default;
+                            Unsafe.Add(ref ip, (nint)i + 2) = default;
+                            Unsafe.Add(ref ip, (nint)i + 3) = default;
+                            Unsafe.Add(ref ip, (nint)i + 4) = default;
+                            Unsafe.Add(ref ip, (nint)i + 5) = default;
+                            Unsafe.Add(ref ip, (nint)i + 6) = default;
+                            Unsafe.Add(ref ip, (nint)i + 7) = default;
+                        } while ((i += 8) < stopLoopAtOffset);
+                    }
+                }
 
-        Write2To3:
-            Debug.Assert(pointerSizeLength >= 2);
+                // Write next 4 elements if needed
+                // Skip this check if "write 4 elements" would've gone down the vectorized code path earlier
 
-            // Write first two and last one.
-            Unsafe.Add(ref ip, 1) = default;
-            Unsafe.Add(ref Unsafe.Add(ref ip, (nint)pointerSizeLength), -1) = default;
+                if (!Vector.IsHardwareAccelerated || Vector<byte>.Count / sizeof(IntPtr) > 4) // JIT turns this into const true or false
+                {
+                    if ((pointerSizeLength & 4) != 0)
+                    {
+                        Unsafe.Add(ref ip, (nint)i + 0) = default;
+                        Unsafe.Add(ref ip, (nint)i + 1) = default;
+                        Unsafe.Add(ref ip, (nint)i + 2) = default;
+                        Unsafe.Add(ref ip, (nint)i + 3) = default;
+                        i += 4;
+                    }
+                }
 
-        Write1:
-            Debug.Assert(pointerSizeLength >= 1);
+                // Write next 2 elements if needed
+                // Skip this check if "write 2 elements" would've gone down the vectorized code path earlier
 
-            // Write only element.
-            ip = default;
+                if (!Vector.IsHardwareAccelerated || Vector<byte>.Count / sizeof(IntPtr) > 2) // JIT turns this into const true or false
+                {
+                    if ((pointerSizeLength & 2) != 0)
+                    {
+                        Unsafe.Add(ref ip, (nint)i + 0) = default;
+                        Unsafe.Add(ref ip, (nint)i + 1) = default;
+                        i += 2;
+                    }
+                }
+
+                // Write final element if needed
+
+                if ((pointerSizeLength & 1) != 0)
+                {
+                    Unsafe.Add(ref ip, (nint)i) = default;
+                }
```

Comment:
```suggestion
                Unsafe.Add(ref ip, (nint)(pointerSizeLength - 1) = default;
```
Can be written at the end directly. Same as https://github.com/dotnet/runtime/pull/51365#discussion_r614840239 applies here.

```diff
-test      dl,1
-je        short M01_L04
-mov       [rcx+rax],0x0
+mov       [rcx+rdx+0FFFF],0x0
```

Unfortunately in the other PR I forgot to look at the scalar path 😢 
https://github.com/dotnet/runtime/blob/ccc25a90082fa6592eeca8e146ff37e1b7d6b79d/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs#L182-L185

Can you please update this? (I don't want to submit a PR for this mini change when you're on it.)